### PR TITLE
Fix keys metrics accounting for CACHE() dictionary with duplicates in the source

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix keys metrics accounting for CACHE() dictionary with duplicates in the source (leads to `DictCacheKeysRequestedMiss` overflows)

Cc: @kitaisreal

### CODE REVIEW

File: src/Dictionaries/CacheDictionary.cpp [hunk 1]
```diff
@@ -609,10 +609,17 @@ void CacheDictionary<dictionary_key_type>::update(CacheDictionaryUpdateUnitPtr<d
...
-                    not_found_keys.erase(fetched_key_from_source);
-                    update_unit_ptr->requested_keys_to_fetched_columns_during_update_index[fetched_key_from_source] = found_keys_size;
-                    found_keys_in_source.emplace_back(fetched_key_from_source);
-                    ++found_keys_size;
+                    /// The underlying source can have duplicates,
+                    /// but if we will process same key twice it will incorrectly account found_keys_size,
+                    /// and also overflows DictCacheKeysRequestedMiss.
+                    ///
+                    /// So only the first row with the same key will be processed.
+                    if (not_found_keys.erase(fetched_key_from_source))
```
Comment (kitaisreal) [inline src/Dictionaries/CacheDictionary.cpp:617]: It probably breaks invariant of insertColumnsForKeys method of ICacheDictionaryStorage, number of columns rows  should equal to number of found keys, and they should be matched.

Comment (kitaisreal) [reply src/Dictionaries/CacheDictionary.cpp:617]: Maybe better to remove found_keys_size here, instead use found_keys_in_source.size(), but at the end of function add such lines
```
size_t found_keys_size = requested_keys_size - not_found_keys.size()
ProfileEvents::increment(ProfileEvents::DictCacheKeysRequestedMiss, not_found_keys);
ProfileEvents::increment(ProfileEvents::DictCacheKeysRequestedFound, found_keys_size);
ProfileEvents::increment(ProfileEvents::DictCacheRequests);
```

Comment (azat) [reply src/Dictionaries/CacheDictionary.cpp]: Good idea, done

**Comment by azat** [comment]
>ClickHouse special build check — 5/6 builds are OK

Looks like CI issue

>Integration tests (asan) — Timeout, fail: 0, passed: 1306, flaky: 2
>Testflows check — failed: 8, passed: 833, other: 100

Unrelated

**Review by kitaisreal — APPROVED** [review]

**Pull request merged**